### PR TITLE
Stop Hot Reload on project unload

### PR DESCRIPTION
**Pull request by tmeschter**
Fixes [AB#1397443](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1397443).

Currently, if the user unloads a project or closes the solution while a Hot Reload session is in progress the session remains, but we now lack any way to interact with it. In some situations this "zombie" session can block changes to later sessions or cause other problems.

The fix here is to end a project's active sessions when the project is unloaded and the `ProjectHotReloadSessionManager` is disposed. Note that we don't wait for the call to `StopSessionAsync` to complete as we just need the session to be stopped sometime soon.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7636)

### CODE REVIEW

**Review by swesonga — APPROVED** [review]
Out of curiosity, why did you choose to use a semaphore with capacity 1 instead C# lock statements in this implementation?

**Comment by tmeschter** [comment]
`await` can't be used within a `lock` block--if control is returned to the caller (e.g. if the awaited `Task` is not complete) the lock will be released.

**Review by BillHiebert — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs [hunk 1]
```diff
@@ -127,6 +129,20 @@ public async Task<bool> TryCreatePendingSessionAsync(IDictionary<string, string>
...
 
+        private async Task OnUnconfiguredProjectUnloadingAsync(object? sender, EventArgs args)
+        {
+            _project.ProjectUnloading -= OnUnconfiguredProjectUnloadingAsync;
+
+            using AsyncSemaphore.Releaser semaphoreRelease = await _semaphore.EnterAsync();
+
+            foreach (HotReloadState sessionState in _activeSessions.Values)
+            {
+                _ = sessionState.Session?.StopSessionAsync(default);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs:140]: Would you want to have any exceptions here logged? If so, you could use the `RunAndForget` extension on `IProjectThreadingService`.

**Review by drewnoakes — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs:140]: You don't need to hook onto Unload, just perform this in your dispose method which will be called on shutdown.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs:140]: Components who's lifetimes are tied to the project composition, shut themselves down in Dispose, implement OnceInitializedOnceDisposeAsync.

Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs:140]: Even better. That's the piece I was missing.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**